### PR TITLE
Add support for saving to and restoring from checkpoints

### PR DESCRIPTION
**Pull request by ramon-garcia**
This pull request contributes support for saving and restoring of variables using tensorflow `SaveV2`/`RestoreV2` ops, and therefore it is compatible with checkpoints created with `tf.keras.Model.save_weights`.

It includes a test case.

### CODE REVIEW

**Comment by ramon-garcia** [comment]
A point for discussion: for my own code, I use anyhow crate to have context in exceptions. Otherwise, one gets an error code without any context.

I didn't add anyhow here to avoid add a new dependency to the project, and to be consistent with the rest of this library. But error messages without context are very difficult to diagnose.

**Comment by adamcrume** [comment]
I don't think `anyhow` is appropriate, because it's intended for application code and forces you to use its error type.  That would be awkward for a library, and switching to it would be a breaking change (since it would change the return type of functions).

I'm definitely in favor of including context in errors.  We can manually add context in error messages; it's just a matter of doing it.  To make it easier, we could add

```
impl Status {
  pub fn context(&self, msg: &str) -> Status {
    ...
  }
}
```

and use it with

```
my_tensorflow_function().map_err(|s| s.context("foo"))?;
```

To make this even more concise, we could do the same thing `anyhow` does and provide an extension trait for `Result`, but if we were to expose it outside this crate, we'd probably want to call the methods something other than `context` so they don't conflict for people using both `tensorflow` and `anyhow`.

Changes like this to `Status` should probably be its own PR, though.

**Comment by ramon-garcia** [comment]
Thank you for your reply. Sounds good to me.

File: src/checkpoint.rs [hunk 1]
```diff
@@ -0,0 +1,370 @@
...
+                nd.add_input(tensor_names);
+                nd.add_input(shape_and_slices);
+                let dtypes = all_variable_ops
+                    .iter()
+                    .map(|v| v.get_attr_type("dtype"))
+                    .collect::<Result<Vec<_>, Status>>()?;
+                nd.set_attr_type_list("dtypes", &dtypes[..])?;
+                nd.finish()?
+            };
+            {
```
Comment (adamcrume) [inline src/checkpoint.rs:139]: This scope doesn't do anything.

File: src/checkpoint.rs [hunk 2]
```diff
@@ -0,0 +1,370 @@
...
+            .shape(var.shape.clone())
+            .build(&mut scope.with_op_name(var.name.as_str()))?;
+        Ok((
+            placeholder.clone(),
+            ops::assign(var.output.clone(), placeholder, scope)?,
+        ))
+    }
+
+    struct MyScopeData {
+        pub scope: Scope,
```
Comment (adamcrume) [inline src/checkpoint.rs:239]: There's no need to make these `pub`.

File: src/checkpoint.rs [hunk 3]
```diff
@@ -0,0 +1,370 @@
...
+
+#[derive(Debug)]
+struct SaveRestoreOps {
+    pub prefix_save: Operation,
+    pub prefix_restore: Operation,
+    pub save_op: Operation,
+    pub restore_op: Operation,
+}
+
+/// Checkpointing and restoring struct
```
Comment (adamcrume) [inline src/checkpoint.rs:14]: This documentation needs to be a little more in-depth.

File: src/checkpoint.rs [hunk 4]
```diff
@@ -0,0 +1,370 @@
+//! This module supports saving and restoring variables using Tensorflow checkpoints in SaveV2 format
```
Comment (adamcrume) [inline src/checkpoint.rs:1]: All the docs need to have punctuation (i.e. end with periods).

**Review by adamcrume — CHANGES_REQUESTED** [review]
Looks great overall, just a few readability concerns.

**Comment by ramon-garcia** [comment]
Thank you for your review. 
In addition, something I am not comfortable with, wouldn't it be better to use an array of variables rather than variable names? When I wrote it, I was concerned about not being able to create a Variable object (for instance, from an imported graph), but it seems to be an independent problem on its own, handled better in its own PR.

File: src/checkpoint.rs [hunk 5]
```diff
@@ -0,0 +1,370 @@
...
+            scope,
+            variables,
+            save_restore_ops: None,
+        }
+    }
+
+    /// Add save and restore ops to the graph
+    fn build_save_ops(&mut self) -> Result<SaveRestoreOps, Status> {
+        let mut all_variable_ops_opt: Option<Vec<Operation>> = None;
+        fn make_all_variable_ops(myself: &mut CheckpointMaker) -> Result<Vec<Operation>, Status> {
```
[the remark this answers is not in the document]
Comment (ramon-garcia) [reply src/checkpoint.rs:39]: The function make_all_variable_ops can be removed if we change from variable names to Variable objects. This is what I am going to implement. Then we can go back if desired.

File: src/checkpoint.rs [hunk 6]
```diff
@@ -0,0 +1,370 @@
+//! This module supports saving and restoring variables using Tensorflow checkpoints in SaveV2 format
+
+use crate::option_insert_result::OptionInsertWithResult;
+use crate::{ops, Operation, Scope, Session, SessionRunArgs, Status, Tensor};
+
+#[derive(Debug)]
+struct SaveRestoreOps {
+    pub prefix_save: Operation,
```
[the remark this answers is not in the document]
Comment (ramon-garcia) [reply src/checkpoint.rs:8]: Done.

File: src/option_insert_result.rs [hunk 7]
```diff
@@ -0,0 +1,18 @@
+// Similar to Option<T>.get_or_insert_with, for a function that returns a result.
+pub trait OptionInsertWithResult<T> {
+    fn get_or_insert_with_result<F, E>(&mut self, f: F) -> Result<&T, E>
```
[the remark this answers is not in the document]
Comment (ramon-garcia) [reply src/option_insert_result.rs:3]: Ok, done

File: src/checkpoint.rs (hunk 2 above)
Comment (ramon-garcia) [reply src/checkpoint.rs:239]: Done

File: src/checkpoint.rs (hunk 3 above)
Comment (ramon-garcia) [reply src/checkpoint.rs:14]: Ok, I added documentation and a small example at the start of the module.

**Comment by ramon-garcia** [comment]
Added to commits, the first one `013a829`. The next one, `2247ed8` changes the specification of variables from strings to Variable objects. 

Best regards.

**Comment by ramon-garcia** [comment]
Hello?

**Review by adamcrume — CHANGES_REQUESTED** [review]

File: src/checkpoint.rs [hunk 8]
```diff
@@ -1,51 +1,72 @@
-//! This module supports saving and restoring variables using Tensorflow checkpoints in SaveV2 format
-
+//! This module supports saving and restoring variables using Tensorflow checkpoints in SaveV2 format.
+//! First, the user creates a [`CheckpointMaker`] attached to an [`Scope`] indicating the list of variables to be saved/restored.
+//! The CheckpointMaker lazily modifies the graph creating the nodes needed for saving/restoring.
+//! When one wants to save/restore from or into a session, one calls the save/restore  methods
+//! # Example
```
[the remark this answers is not in the document]
Comment (ramon-garcia) [reply src/checkpoint.rs]: I hope it is now OK

File: src/checkpoint.rs [hunk 9]
```diff
@@ -1,51 +1,72 @@
...
+/*    fn make_all_variable_ops(&mut self) -> Result<Vec<Operation>, Status> {
+        let graph = self.scope.graph();
+        Ok(self
+            .variables
+            .iter()
+            .map(|v: &String| -> Result<Operation, Status> {
+                Ok(graph.operation_by_name_required(v.as_str())?.clone())
+            })
+            .collect::<Result<Vec<_>, Status>>()?)
+    }*/
```
[the remark this answers is not in the document]
Comment (ramon-garcia) [reply src/checkpoint.rs]: Done

**Comment by ramon-garcia** [comment]
ping.

**Comment by adamcrume** [comment]
Sorry for the delay, and thank you!

**Pull request merged**